controlplane/kubeadm/controllers/controller.go

```diff
@@ -200,6 +188,12 @@ func (r *KubeadmControlPlaneReconciler) Reconcile(req ctrl.Request) (res ctrl.Re
 		}
 	}()
 
+	// Add finalizer first if not exist to avoid the race condition between init and delete
```

Comment:
If I'm not mistaken, move this add finalizer check after the above defer() block will bring back the issue of #https://github.com/kubernetes-sigs/cluster-api/issues/3198

Because within the defer() it is trying to get target cluster status which will take 30 seconds to timeout.

Reply:
Emm, this means that we have to check what has been done in defer. 
I've tried to decompose the origin big defer into two small defers : the first defer does only simple patch operation and is fast enough, and the second defer handles more complex and potentially slow operations. Then we can check and add the finalizer between the two defers.

Reply:
not quite sure about this. 

first, defered funcs are called last-in-first-out order.  secondly, after status updates/reconciles, need a second patch() to persist the status. 

Also, I think the issue only exists in kcp controller, other controllers does not have the time consuming call to get remote cluster status.  So if we keep your original implementation, then one possible fix maybe just add a check within updateStatus, try to get remote cluster status only when len(ownedMachines) > 0.

Reply:
Yes, defer funcs are called last-in-first-out order. So the first defer will always be executed last, then the status after updates/reconciles can be persisted. The advantage of splitting is that we don't have to handle the details of defer funcs separately. 

I want to wait for more comments about this approach.  And then chose the better one.
/cc @vincepri @detiber

Reply:
got it. You are right. This should work. Maybe add some comments.

Reply:
For KCP specifically it should be fine patching right away instead of reusing the main defer, let's only make sure to add a comment explaining the reasoning and we should we good to go

Reply:
OK, let's revert it back and just add some comments.